server/controllers/handleQuestItemClicked.js

```diff
@@ -70,47 +69,70 @@ export const handleQuestItemClicked = async (req, res) => {
+              currentStreak: 1,
+              lastCollectedDate: currentDate,
+              longestStreak: 1,
+              total: 1,
+              totalCollectedToday: 1,
+            },
             [`profileMapper.${profileId}`]: username,
           }),
         );
-        numberCollected = 1;
-      } else if (!itemsCollectedByUser?.[profileId]?.[dateKey]) {
-        // first time user has interacted with Quest on this day
+      } else {
+        let currentStreak = itemsCollectedByUser[profileId].currentStreak || 1;
+        let total = itemsCollectedByUser[profileId].total ? itemsCollectedByUser[profileId].total + 1 : 1;
+        let totalCollectedToday = itemsCollectedByUser[profileId].totalCollectedToday || 1;
+        const lastCollectedDate = itemsCollectedByUser[profileId].lastCollectedDate;
+
+        let longestStreak = itemsCollectedByUser[profileId].longestStreak;
+        if (!longestStreak) longestStreak = getLongestStreak(itemsCollectedByUser[profileId]);
+
+        if (lastCollectedDate) {
+          const differenceInDays = getDifferenceInDays(lastCollectedDate, currentDate);
+          if (differenceInDays === 0) {
+            totalCollectedToday = totalCollectedToday + 1;
+            numberCollectedToday = totalCollectedToday + 1;
+          } else if (differenceInDays === 1) {
+            currentStreak = currentStreak + 1;
+          }
+        }
+
+        if (currentStreak + 1 > longestStreak) longestStreak = currentStreak;
+
+        if (totalCollectedToday + 1 === numberAllowedToCollect) analytics.push("dayCompletedCount");
+
         promises.push(
           world.updateDataObject({
-            [`keyAssets.${keyAssetId}.itemsCollectedByUser.${profileId}.${dateKey}`]: { count: 1 },
-            [`keyAssets.${keyAssetId}.itemsCollectedByUser.${profileId}.total`]:
-              itemsCollectedByUser[profileId].total + 1,
+            [`keyAssets.${keyAssetId}.itemsCollectedByUser.${profileId}`]: {
+              currentStreak,
+              lastCollectedDate: currentDate,
+              longestStreak,
+              total,
+              totalCollectedToday,
+            },
           }),
         );
-        numberCollected = 1;
-      } else {
-        promises.push(
-          world.incrementDataObjectValue(
-            [`keyAssets.${keyAssetId}.itemsCollectedByUser.${profileId}.${dateKey}.count`],
-            1,
-          ),
-        );
-        promises.push(
-          world.incrementDataObjectValue([`keyAssets.${keyAssetId}.itemsCollectedByUser.${profileId}.total`], 1),
-        );
-        numberCollected = itemsCollectedByUser[profileId][dateKey].count + 1;
       }
 
       const visitor = await getVisitor({
         credentials,
         urlSlug,
         visitorId,
       });
-      if (itemsCollectedByUser[profileId].total + 1 === 5) {
-        const test = await visitor.grantExpression({ name: "quest_1" });
+
+      if ([30, 50, 75].includes(itemsCollectedByUser[profileId].total + 1)) {
```

Comment:
`itemsCollectedByUser[profileId].total + 1`
calculated multiple times. I have no issues with that though, just pointing out